pkg/kubelet/status/status_manager.go

```diff
@@ -829,32 +837,58 @@ func (m *manager) syncBatch(all bool) int {
 		}
 	}()
 
+	podInfos := make(map[string]v1.Pod)
+	updatedStatusesCount := len(updatedStatuses)
+	if m.nodeName != "" && updatedStatusesCount >= defaultMinListPodCount {
+		podOpts := metav1.ListOptions{FieldSelector: fields.OneTermEqualSelector("spec.nodeName", m.nodeName).String()}
+		pods, err := m.kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), podOpts)
+		if err != nil {
+			klog.InfoS("Failed to list pods for node", "node", m.nodeName, "err", err)
+		} else if pods != nil && len(pods.Items) > 0 {
+			for index := range pods.Items {
+				podInfo := pods.Items[index]
+				key := fmt.Sprintf("%s_%s", podInfo.Name, podInfo.Namespace)
+				podInfos[key] = podInfo
+			}
+		}
+	}
+
+	// TODO: use ParallelizeUntil to do same syncPod
 	for _, update := range updatedStatuses {
 		klog.V(5).InfoS("Sync pod status", "podUID", update.podUID, "statusUID", update.statusUID, "version", update.status.version)
-		m.syncPod(update.podUID, update.status)
+		key := fmt.Sprintf("%s_%s", update.status.podName, update.status.podNamespace)
+		podInfo, ok := podInfos[key]
+		if ok {
+			m.syncPod(update.podUID, update.status, &podInfo)
+		} else {
+			m.syncPod(update.podUID, update.status, nil)
+		}
 	}
 
 	return len(updatedStatuses)
 }
 
 // syncPod syncs the given status with the API server. The caller must not hold the status lock.
-func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
-	// TODO: make me easier to express from client code
-	pod, err := m.kubeClient.CoreV1().Pods(status.podNamespace).Get(context.TODO(), status.podName, metav1.GetOptions{})
-	if errors.IsNotFound(err) {
-		klog.V(3).InfoS("Pod does not exist on the server",
-			"podUID", uid,
-			"pod", klog.KRef(status.podNamespace, status.podName))
-		// If the Pod is deleted the status will be cleared in
-		// RemoveOrphanedStatuses, so we just ignore the update here.
-		return
-	}
-	if err != nil {
-		klog.InfoS("Failed to get status for pod",
-			"podUID", uid,
-			"pod", klog.KRef(status.podNamespace, status.podName),
-			"err", err)
-		return
+func (m *manager) syncPod(uid types.UID, status versionedPodStatus, pod *v1.Pod) {
+	if pod == nil {
+		// TODO: make me easier to express from client code
+		var err error
+		pod, err = m.kubeClient.CoreV1().Pods(status.podNamespace).Get(context.TODO(), status.podName, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			klog.V(3).InfoS("Pod does not exist on the server",
+				"podUID", uid,
+				"pod", klog.KRef(status.podNamespace, status.podName))
+			// If the Pod is deleted the status will be cleared in
+			// RemoveOrphanedStatuses, so we just ignore the update here.
+			return
+		}
+		if err != nil {
+			klog.InfoS("Failed to get status for pod",
+				"podUID", uid,
+				"pod", klog.KRef(status.podNamespace, status.podName),
+				"err", err)
+			return
+		}
```

Comment:
Could we avoid reading the old status entirely by constructing the status patch directly, rather than merging the old status and then effectively unmerging it to generate the patch (lines 905-907 below)? Or is the patch still conditional on the old status?